### PR TITLE
feat: Support Trimmed Function Names

### PR DESCRIPTION
**Pull request by mitsuhiko**
This PR adds support for the new `function_name` attribute and uses it for the newest grouping strategy. The reason for this is that the existing function attribute contains a lot of extra information that is often not useful when it gets too long and obfuscates the information we actually are about.

Why do we do this? Because for many native customers with heavy C++ templated code they cannot actually currently tell what the name of a function in the stacktrace is. This PR improves this greatly.

As an example STL in C++ can produce function names which are way longer than the 255 character limit we allow but the most useful information is in the trimmed away part. The purpose of `function_name` is that the native processor can pre-trim the function before storing it. This way even if the function before storing is beyond 255 characters the useful information can be extracted beforehand.

For the grouping this information is also picked up by the newest newstyle grouping but not by the older one which uses the function name and the existing late trimming logic instead.

## UI Changes

trimming mov

<details><summary>Screenshots</summary>
<br>

**Trimmed functions**

image

**Untrimmed functions**

image

</details>

## Open Questions

* Do we always want to store `function_name`? We do currently but we could also only store it if it's different to `function`. I think I might prefer that because it keeps the size down and we need to handle func == function_name in the UI anyways.

### CODE REVIEW

**Comment by mitsuhiko** [comment]
@benvinegar @billyvg I assigned you as reviewers because this is product facing. This change currently only affects native but I want to make it available to all customers if we go forward with this because the current behavior is objectively worse than this.

I can show you some stacktraces that are effectively unreadable because the actual method name cannot be seen since it's past the 256 character limit.

**Comment by mitsuhiko** [comment]
This PR proposes to store the trimmed function name for native in `function_name` rather than `function` which it leaves unchanged. However I also realized that we forward this value into snuba (`stack.function`) which makes me wonder if we want to either do this the other way round or change snuba to get `function_name` if set.

The ingestion for functions from cocoa puts the DWARF symbol into `function` and during processing if trimming results in something we put it into `function_name` after trimming. The same thing happens for minidumps but there the situation is easier as all values are produced on the server.

We could alternatively do what @jan-auer proposed which is to change the internal minidump processing to put the raw function into `function_raw` and to store the trimmed function name in `function_name`. If we do this we need to make sure that the legacy grouping algorithm picks up `function_raw` before `function` on grouping and this also then causes all new events to show up with those trimmed functions in snuba and the UI. I'm starting to lean in that direction because while it's the more invasive change, likely we want trimmed functions.

@eurica @benvinegar Do we have any preferences from a product situation? I think trimmed functions for native are in every way superior as the default view and I would also prefer having them in snuba as a user if I want to use them in discover.

**Comment by mitsuhiko** [comment]
@jan-auer if we go with `function_raw` over `function_name` I think we could make it work with the following logic:

* the legacy grouping strategy prefers `function_raw` over `function`. This solves grouping concerns
* all minidump and native processing code emits `function_raw` for symbolication and puts the trimmed function into `function`
* for native events that have a function name incoming (cocoa) we treat the incoming function name in the frame as raw function instead and trim it. If they are different before and after trimming we put the trimmed into `function` and the original value into `function_raw`, otherwise we leave it unchanged.
* This causes UI, plugins and snuba to pick up the trimmed function consistently.

Doing so will change data forwarders and the raw stacktrace feature as well as the JSON payload after processing however since we have both values grouping will not be affected and the UI functionality from this PR would still work as before (toggle on click)

**Comment by benvinegar** [comment]
I like the UX. The intent (vs `function`) makes sense.

File: src/sentry/stacktraces/functions.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    """
+    if get_behavior_family_for_platform(platform) != 'native':
+        return function
+    if function in ('<redacted>', '<unknown>'):
+        return function
+
+    original_function = function
+    function = function.strip()
+
+    # Ensure we don't operated on objc functions
```
Comment (markstory) [inline src/sentry/stacktraces/functions.py]:
```suggestion
    # Ensure we don't operate on objc functions
```
:bike:

File: src/sentry/static/sentry/app/components/events/interfaces/frame.jsx [hunk 2]
```diff
@@ -25,6 +25,44 @@ export function trimPackage(pkg) {
...
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      fullFunction: false,
+    };
+  }
+
+  toggle(event) {
```
Comment (markstory) [inline src/sentry/static/sentry/app/components/events/interfaces/frame.jsx]:
```suggestion
  toggle = (event) => {
```

Using a fat arrow function will let you skip the `.bind(this)` in the render() call.

File: src/sentry/static/sentry/app/components/events/interfaces/frame.jsx [hunk 3]
```diff
@@ -25,6 +25,44 @@ export function trimPackage(pkg) {
...
+    const canToggle = func && funcName && func !== funcName;
+
+    if (!canToggle) {
+      return <code {...props}>{funcName || func || '<unknown>'}</code>;
+    }
+
+    const current = this.state.fullFunction ? func : funcName;
+    return (
+      <code {...props}>
+        <a onClick={this.toggle.bind(this)}>{current || '<unknown>'}</a>
```
Comment (markstory) [inline src/sentry/static/sentry/app/components/events/interfaces/frame.jsx]:
```suggestion
        <a onClick={this.toggle}>{current || '<unknown>'}</a>
```

File: src/sentry/static/sentry/app/components/events/interfaces/frame.jsx [hunk 4]
```diff
@@ -25,6 +25,44 @@ export function trimPackage(pkg) {
...
+    super(props);
+    this.state = {
+      fullFunction: false,
+    };
+  }
+
+  toggle = event => {
+    event.stopPropagation();
+    this.setState({
+      fullFunction: !this.state.fullFunction,
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/events/interfaces/frame.jsx]: Let's use setState callback when referencing prev state. See: https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-access-state-in-setstate.md

**Review by billyvg — COMMENTED** [review]
Looks good, have a few UX questions:

* Why is there still an ellipsis when it's expanded?
image

* Can we show an indicator (e.g. ellipsis) when `function` !== `functionName`?

**Comment by mitsuhiko** [comment]
@billyvg the ellipsis is not intentional, it's because the original function name was too long so it was trimmed during processing. The `...` is our standard marker to indicate that not all information is available.

Some of these function names can be more than 16k characters and we trim to 255.

> Can we show an indicator (e.g. ellipsis) when function !== functionName?

Any ideas how? Currently i indicate it by the underline over effect.

**Comment by billyvg** [comment]
@mitsuhiko Don't have any good ideas since we use `...` and already have a expand/collapse

**Comment by mitsuhiko** [comment]
@billyvg i'm not 100% happy with the current solution because i had to prevent the bubbling as otherwise the expand frame logic also kicks in.

**Comment by eurica** [comment]
If @benvinegar is happy, I'm happy.  Looks like a good change for those long names.

**Comment by mitsuhiko** [comment]
As per last conversation we're going with #12990 instead.

**Pull request closed without merging**